[KumChaiYin] iP

KumChaiYin opened the pull request:
# AiChan 🌟 
> "Your mind is for having ideas, not holding them." -David Allen ([source](https://dansilvestre.com/productivity-quotes/))

AiChan frees your mind of having to remember things you need to do. It's, 
- text-based
- ~~EASY~~ _VERY_ EASY to learn
- SUPER FAST to use

All you need to do is, 
1. download it from [here](https://github.com/KumChaiYin/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you

And it is **FREE**! 🔥 

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon) ✨ 
---
If you are Java programmer, you can use it to practice Java too. Here's the <code style="color: red;">main</code> method:
 ```java
    public static void main(String[] args) {
        new AiChan("data/tasks.txt").run();
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

KumChaiYin committed:
Rename, Greet, Exit

KumChaiYin committed:
lvl1 echo

KumChaiYin committed:
lvl2 Add, List

KumChaiYin committed:
Level-3

KumChaiYin committed:
Level-4

KumChaiYin committed:
A-TextUiTesting

KumChaiYin committed:
Level-5

KumChaiYin committed:
Level-6

KumChaiYin committed:
A-Enums

KumChaiYin committed:
update testing

KumChaiYin committed:
A-MoreOOP

KumChaiYin committed:
A-Packages

KumChaiYin committed:
Level-7

KumChaiYin committed:
Merge branch 'branch-Level-7'
no fast-forward merge branch-Level-7

KumChaiYin committed:
branch lvl8

KumChaiYin committed:
Merge branch 'branch-Level-8'

KumChaiYin committed:
Merge remote-tracking branch 'origin/add-gradle-support'
merge add-gradle-support.

KumChaiYin committed:
A-JUnit

KumChaiYin committed:
A-Jar

KumChaiYin committed:
branch-JavaDoc

KumChaiYin committed:
branch-CodingStandard

KumChaiYin committed:
branch lvl9

KumChaiYin committed:
Merge branch 'branch-A-JavaDoc'
First merge of parallel branches.

KumChaiYin committed:
Resolved merge conflicts

KumChaiYin committed:
Resolved merge conflicts 2

AndrewJanong commented on src/main/java/aichan/task/Task.java:
```diff
@@ -0,0 +1,100 @@
+     * @return The corresponding task.
+     * @throws AiChanException If the line has incorrect format.
+     */
+    public static Task stringToTask(String str) throws AiChanException {
```
Perhaps you could use verbs to name the method? Maybe something like `getTaskFromCommand(String str)`

```suggestion
    public static Task getTaskFromCommand(String str) throws AiChanException {
```

AndrewJanong commented on src/main/java/aichan/task/Task.java:
```diff
@@ -0,0 +1,100 @@
+     * @throws AiChanException If the line has incorrect format.
+     */
+    public static Task stringToTask(String str) throws AiChanException {
+        char type = str.charAt(0);
+        boolean isMark = (str.charAt(4) == '1');
+        if (str.length() < 8) {
+            throw new AiChanException("incorrect format in the file");
+        }
+        String description = str.substring(8);
+        if (type == 'T') {
+            Task t = new ToDo(description);
+            if (isMark) t.mark();
+            return t;
+        } else if (type == 'D') {
+            String[] arr = description.split(" \\| ");
+            Task t = new Deadline(arr);
+            if (isMark) t.mark();
+            return t;
+        } else if (type == 'E') {
+            String[] arr = description.split(" \\| | - ");
+            Task t = new Event(arr);
+            if (isMark) t.mark();
+            return t;
+        } else {
+            return null;
+        }
+    }
```
* I'm not sure I like how cramped the code is. Maybe can add some **line breaks** to improve readabiliity?
* Not using {} for if statements does not follow the coding standard.

```suggestion
        char type = str.charAt(0);
        boolean isMark = (str.charAt(4) == '1');
        
        if (str.length() < 8) {
            throw new AiChanException("incorrect format in the file");
        }
        
        String description = str.substring(8);
        if (type == 'T') {
            Task t = new ToDo(description);
            if (isMark) {
                t.mark();
            }
            return t;
        } else if (type == 'D') {
            String[] arr = description.split(" \\| ");
            Task t = new Deadline(arr);
            if (isMark) {
                t.mark();
            }
            return t;
        } else if (type == 'E') {
            String[] arr = description.split(" \\| | - ");
            Task t = new Event(arr);
            if (isMark) {
                t.mark();
            }
            return t;
        } else {
            return null;
        }
    }

```

AndrewJanong commented on src/main/java/aichan/Storage.java:
```diff
@@ -0,0 +1,86 @@
+     * @param tasks TaskList contains current tasks.
+     * @throws AiChanException If an error occurs when saving tasks.
+     */
+    public void save(TaskList tasks) throws AiChanException {
```
Maybe can give a more descriptive function name?

```suggestion
    public void saveTasks(TaskList tasks) throws AiChanException {
```

AndrewJanong commented on src/main/java/aichan/Storage.java:
```diff
@@ -0,0 +1,86 @@
+     * @return ArrayList which contains the tasks.
+     * @throws AiChanException If an error occurs when loading tasks.
+     */
+    public ArrayList<Task> load() throws AiChanException {
```
Same issue as before, function name could maybe be more descriptive.

```suggestion
    public ArrayList<Task> loadTasks() throws AiChanException {
```

AndrewJanong commented on src/main/java/aichan/Storage.java:
```diff
@@ -0,0 +1,86 @@
+        ArrayList<Task> arrTask = new ArrayList<>();
+
+        try {
+            File file = new File(this.filePath);
+            if (!file.exists()) {
+                return arrTask;
+            }
+            Scanner scn = new Scanner(file);
+            while (scn.hasNextLine()) {
+                String line = scn.nextLine();
+                if (line.isEmpty()) {
+                    break;
+                }
+                Task task = Task.stringToTask(line);
+                if (task != null) {
+                    arrTask.add(task);
+                }
+            }
+            scn.close();
```
... I noticed the same issue in several other places.

```suggestion
            File file = new File(this.filePath);
            
            if (!file.exists()) {
                return arrTask;
            }
            
            Scanner scn = new Scanner(file);
                   
            while (scn.hasNextLine()) {
                String line = scn.nextLine();
                if (line.isEmpty()) {
                    break;
                }
                
                Task task = Task.stringToTask(line);
                if (task != null) {
                    arrTask.add(task);
                }
            }
            
            scn.close();

```

AndrewJanong commented on src/main/java/aichan/AiChan.java:
```diff
@@ -0,0 +1,64 @@
+        }
+    }
+
+    /**
+     * Runs the main loop of application AiChan.
+     * Greets user, receives commands from user and executes it.
+     * Terminates when user enter "bye".
+     */
```
I like how you describe your methods, short but clear. 👍

AndrewJanong reviewed (commented):
## Overall Review
Overall I found your code easy to read for the most part, but there are some issues I came across:
* **your code is too cramped** in some places. Adding some line breaks will improve code readability. 
* **method names are undescriptive** is another issue I see often. Using verbs to name methods would be better.

KumChaiYin committed:
A-CheckStyle

KumChaiYin committed:
follow coding standard

KumChaiYin committed:
Merge tag 'branch-A-CheckStyle'
Follow Java coding standard

Modifying some method names and adding some line breaks.

KumChaiYin committed:
add a GUI to the chatbot

KumChaiYin committed:
Merge branch 'branch-Level-10'
Add a GUI to the chatbot.

KumChaiYin committed:
use Java varargs in branch

KumChaiYin committed:
Merge branch 'branch-A-Varargs'
Use Java varargs feature in FindCommand class

Jweewee commented on src/main/java/aichan/Parser.java:
```diff
@@ -20,22 +20,25 @@ public static Command parse(String input) throws AiChanException {
+        } else if (commandType.equals(ActionType.DELETE.toString())) {
             return new DeleteCommand(strToNum(checkLen(inputs)));
+        } else if (commandType.equals(ActionType.FIND.toString())) {
+            return new FindCommand(checkLen(inputs));
```
LGTM, namimng convention of commands are descriptive and nouns

Jweewee commented on src/main/java/aichan/ActionType.java:
```diff
@@ -8,7 +8,8 @@ public enum ActionType {
     LIST("list"),
-    BYE("bye");
+    BYE("bye"),
+    FIND("find");
```
LGTM, correct naming convention

Jweewee reviewed (commented):
LGTM

KumChaiYin committed:
Use assert feature

To check that the array argument passed to the Deadline and Event
constructor have enough length.

Check that the length of String array:
* is bigger than 2 in Deadline constructor.
* is bigger than 3 in Event constructor.

KumChaiYin committed:
Improve code quality

Correct some error-prone practices and improve readability.

AiChan: delete dead code.
List Command: change variable name i to taskId
ExitCommand:
* remove an obvious comment
* avoid a magic String in method execute

KumChaiYin committed:
Use Streams feature

Remove the nested loop in FindCommand class.

Use Stream to filter the empty string inside keywords array.
Use Stream to check whether a task name contains keyword.

KumChaiYin committed:
Merge pull request #1 from KumChaiYin/branch-A-Assertions

Use assert feature

KumChaiYin committed:
Merge pull request #2 from KumChaiYin/branch-A-CodeQuality

code quality

KumChaiYin committed:
Merge branch 'master' into branch-A-CodeQuality

KumChaiYin committed:
Merge pull request #3 from KumChaiYin/branch-A-Streams

Use Streams feature

KumChaiYin committed:
Merge branch 'master' into branch-A-Streams

KumChaiYin committed:
Add extension C-BetterSearch C-MassOps

Allows user to find task which contains partial word in keywords.
Allow user to delete all marked tasks with the command "deleteMarked".

KumChaiYin committed:
Merge branch 'branch-BCD-Extension'

KumChaiYin committed:
Add user guide

KumChaiYin committed:
Adjust formatting

KumChaiYin committed:
Adjust table

KumChaiYin committed:
improve code quality

KumChaiYin committed:
debug gradle

KumChaiYin committed:
Merge branch 'branch-GUI'

KumChaiYin committed:
debug storage
